### PR TITLE
Improve FSDP LoRA Memory Usage

### PR DESCRIPTION
**Pull request by awgu**
**Overview**
This PR improves the FSDP LoRA fine-tuning memory usage and fixes some typos.
- We enable the FSDP rate limiter (which was intentionally disabled for some reason). This decreases reserved memory by 5.31 GB and active memory by 5.66 GB on 4 GPUs.
- We do not keep `logits` (as well as `past_key_values`, `hidden_states`, and `attentions`) alive after forward / through backward by avoiding any references to those tensors via `loss = model(*batch).loss` instead of `outputs = model(*batch); loss = outputs.loss`. This further decreases reserved memory by 8.00 GB and active memory by 4.98 GB on 4 GPUs. (thanks @albanD for the help!)
- Overall, this **decreases peak reserved memory by 13.31 GB and peak active memory by 10.64 GB** on 4 GPUs for pure bf16 training. In the end, this only uses 15.22 GB peak active memory and 19.36 GB peak reserved memory.

**Test Plan**
```
torchrun --nnodes 1 --nproc_per_node 4  llama_finetuning.py --enable_fsdp --use_peft --peft_method lora --model_name llama-2-7b-hf/ --pure_bf16 --output_dir output_dir
```

For my AWS setup with 4 40 GB A100 GPUs:
- With `limit_all_gathers=False`:
    - Peak active: 25.86 GB
    - Peak reserved: 32.67 GB
- With `limit_all_gathers=True`:
    - Peak active: 20.20 GB
    - Peak reserved: 27.36 GB
- With `limit_all_gathers=True` and not keeping logits alive:
    - Peak active: 15.22 GB
    - Peak reserved: 19.36 GB

```
mem_stats = torch.cuda.memory_stats()
peak_active_gb = mem_stats["active_bytes.all.peak"] / (1024 ** 3)
peak_reserved_gb = mem_stats["reserved_bytes.all.peak"] / (1024 ** 3)
print(f"peak active: {peak_active_gb} GB | peak reserved: {peak_reserved_gb} GB") 
```

### CODE REVIEW

**Comment by awgu** [comment]
logits
This memory snapshot shows the large `logits` tensor that was being held onto through backward.

**Review by HamidShojanazeri — APPROVED** [review]

**Review by chauhang — APPROVED** [review]
